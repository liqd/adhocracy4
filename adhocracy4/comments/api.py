from django.conf import settings

from django_filters import rest_framework as filters
from rest_framework import mixins, viewsets
from rest_framework.response import Response

from adhocracy4.api.mixins import ContentTypeMixin
from adhocracy4.api.permissions import ViewSetRulesPermission

from .models import Comment
from .serializers import ThreadSerializer


class CommentViewSet(mixins.CreateModelMixin,
                     mixins.RetrieveModelMixin,
                     mixins.UpdateModelMixin,
                     mixins.DestroyModelMixin,
                     ContentTypeMixin,
                     viewsets.GenericViewSet):

    queryset = Comment.objects.all().order_by('-created')
    serializer_class = ThreadSerializer
    permission_classes = (ViewSetRulesPermission,)
    filter_backends = (filters.DjangoFilterBackend,)
    filter_fields = ('object_pk', 'content_type')
    content_type_filter = settings.A4_COMMENTABLES

    def perform_create(self, serializer):
        serializer.save(
            content_object=self.content_object,
            creator=self.request.user
        )

    def get_permission_object(self):
        return self.content_object

    @property
    def rules_method_map(self):
        return ViewSetRulesPermission.default_rules_method_map._replace(
            POST='{app_label}.comment_{model}'.format(
                app_label=self.content_type.app_label,
                model=self.content_type.model
            )
        )

    def destroy(self, request, *args, **kwargs):
        comment = self.get_object()
        if self.request.user == comment.creator:
            comment.is_removed = True
        else:
            comment.is_censored = True
        comment.save()
        serializer = self.get_serializer(comment)
        return Response(serializer.data)
