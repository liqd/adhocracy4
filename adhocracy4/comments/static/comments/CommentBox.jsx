var CommentList = require('./CommentList')
var CommentForm = require('./CommentForm')
var CommentCategoryForm = require('./CommentCategoryForm')

var api = require('../../../static/api')

var React = require('react')
var update = require('immutability-helper')
var PropTypes = require('prop-types')
var django = require('django')

class CommentBox extends React.Component {
  constructor (props) {
    super(props)

    this.state = {
      comments: this.props.comments
    }
  }

  updateStateComment (index, parentIndex, updatedComment) {
    var comments = this.state.comments
    var diff = {}
    if (typeof parentIndex !== 'undefined') {
      diff[parentIndex] = { child_comments: {} }
      diff[parentIndex].child_comments[index] = { $merge: updatedComment }
    } else {
      diff[index] = { $merge: updatedComment }
    }
    comments = update(comments, diff)
    this.setState({ comments: comments })
  }

  handleCommentSubmit (comment, parentIndex) {
    return api.comments.add(comment)
      .done(comment => {
        var comments = this.state.comments
        var diff = {}
        if (typeof parentIndex !== 'undefined') {
          diff[parentIndex] = { child_comments: { $push: [ comment ] } }
        } else {
          diff = { $unshift: [ comment ] }
        }
        this.setState({
          comments: update(comments, diff)
        })

        if (typeof parentIndex !== 'undefined') {
          this.updateStateComment(
            parentIndex,
            undefined,
            {
              replyError: false,
              errorMessage: undefined
            })
        } else {
          this.setState({
            error: false,
            errorMessage: undefined
          })
        }
      })
      .fail((xhr, status, err) => {
        var errorMessage = xhr.responseJSON.comment[0]
        if (typeof parentIndex !== 'undefined') {
          this.updateStateComment(
            parentIndex,
            undefined, {
              replyError: true,
              errorMessage: errorMessage
            })
        } else {
          this.setState({
            error: true,
            errorMessage: errorMessage
          })
        }
      })
  }

  handleCommentModify (modifiedComment, index, parentIndex) {
    var comments = this.state.comments
    var comment = comments[index]
    if (typeof parentIndex !== 'undefined') {
      comment = comments[parentIndex].child_comments[index]
    }

    return api.comments.change(modifiedComment, comment.id)
      .done(changed => {
        this.updateStateComment(index, parentIndex, changed)
        this.updateStateComment(
          index,
          parentIndex, {
            editError: false,
            errorMessage: undefined
          }
        )
      })
      .fail((xhr, status, err) => {
        var errorMessage = xhr.responseJSON.comment[0]
        this.updateStateComment(
          index,
          parentIndex,
          {
            editError: true,
            errorMessage: errorMessage
          })
      })
  }

  handleCommentDelete (index, parentIndex) {
    var comments = this.state.comments
    var comment = comments[index]
    if (typeof parentIndex !== 'undefined') {
      comment = comments[parentIndex].child_comments[index]
    }

    var data = {
      urlReplaces: {
        contentTypeId: comment.content_type,
        objectPk: comment.object_pk
      }
    }
    return api.comments.delete(data, comment.id)
      .done(this.updateStateComment.bind(this, index, parentIndex))
  }

  hideNewError () {
    this.setState({
      error: false,
      errorMessage: undefined
    })
  }

  hideReplyError (index, parentIndex) {
    this.updateStateComment(
      index,
      parentIndex,
      {
        replyError: false,
        errorMessage: undefined
      }
    )
  }

  hideEditError (index, parentIndex) {
    this.updateStateComment(
      index,
      parentIndex,
      {
        editError: false,
        errorMessage: undefined
      }
    )
  }

  getChildContext () {
    return {
      isAuthenticated: this.props.isAuthenticated,
      isModerator: this.props.isModerator,
      comments_contenttype: this.props.comments_contenttype,
      user_name: this.props.user_name
    }
  }

  getCommentForm () {
    if (this.props.commentCategoryChoices) {
      return (
        <CommentCategoryForm subjectType={this.props.subjectType} subjectId={this.props.subjectId}
          onCommentSubmit={this.handleCommentSubmit.bind(this)} placeholder={django.gettext('Your comment here')}
          rows="5" isReadOnly={this.props.isReadOnly} error={this.state.error} errorMessage={this.state.errorMessage} handleErrorClick={this.hideNewError.bind(this)}
          commentCategoryChoices={this.props.commentCategoryChoices} />
      )
    } else {
      return (
        <CommentForm subjectType={this.props.subjectType} subjectId={this.props.subjectId}
          onCommentSubmit={this.handleCommentSubmit.bind(this)} placeholder={django.gettext('Your comment here')}
          rows="5" isReadOnly={this.props.isReadOnly} error={this.state.error} errorMessage={this.state.errorMessage} handleErrorClick={this.hideNewError.bind(this)} />
      )
    }
  }

  render () {
    return (
      <div>
        <div className="black-divider">{this.state.comments.length + ' ' + django.ngettext('comment', 'comments', this.state.comments.length)}</div>
        <div className="comment-box">
          {this.getCommentForm()}
          <div className="comment-list">
            <CommentList
              comments={this.state.comments}
              handleCommentDelete={this.handleCommentDelete.bind(this)}
              handleCommentSubmit={this.handleCommentSubmit.bind(this)}
              handleCommentModify={this.handleCommentModify.bind(this)}
              isReadOnly={this.props.isReadOnly}
              handleReplyErrorClick={this.hideReplyError.bind(this)}
              handleEditErrorClick={this.hideEditError.bind(this)}
            />
          </div>
        </div>
      </div>
    )
  }
}

CommentBox.childContextTypes = {
  isAuthenticated: PropTypes.bool,
  isModerator: PropTypes.bool,
  comments_contenttype: PropTypes.number,
  user_name: PropTypes.string
}

module.exports = CommentBox
