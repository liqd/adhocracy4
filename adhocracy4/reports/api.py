from rest_framework import mixins
from rest_framework import permissions
from rest_framework import viewsets

from . import emails
from .models import Report
from .serializers import ReportSerializer


class ReportViewSet(mixins.CreateModelMixin,
                    viewsets.GenericViewSet):

    serializer_class = ReportSerializer
    queryset = Report.objects.all()
    permission_classes = (permissions.IsAuthenticated,)

    def perform_create(self, serializer):
        report = serializer.save(creator=self.request.user)
        emails.ReportModeratorEmail.send(report)
