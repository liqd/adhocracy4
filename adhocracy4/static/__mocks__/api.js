let comments = null
let following = null

const api = {
  comments: {
    get: jest.fn(() => {
      const instance = {
        done: (fn) => {
          if (comments !== null) {
            fn(comments)
          }
          return instance
        },
        fail: (fn) => {
          if (comments === null) {
            fn({ status: 400 })
          }
          return instance
        }
      }
      return instance
    }),
    setComments: (value) => {
      comments = value
    }
  },
  follow: {
    get: jest.fn(() => {
      const instance = {
        done: (fn) => {
          if (following !== null) {
            fn(following)
          }
          return instance
        },
        fail: (fn) => {
          if (following === null) {
            fn({ status: 400 })
          }
          return instance
        }
      }
      return instance
    }),
    change: jest.fn((enabled) => {
      following = { enabled }
      const instance = {
        done: (fn) => {
          if (following !== null) {
            fn(following)
          }
          return instance
        },
        fail: (fn) => {
          if (following === null) {
            fn({ status: 400 })
          }
          return instance
        }
      }
      return instance
    }),
    setFollowing: (value) => {
      following = value
    }
  }
}

module.exports = api
