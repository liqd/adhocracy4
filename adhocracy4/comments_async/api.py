from django.conf import settings
from django.contrib.contenttypes.models import ContentType
from django_filters import rest_framework as filters
from rest_framework import mixins
from rest_framework import status
from rest_framework import viewsets
from rest_framework.pagination import PageNumberPagination
from rest_framework.response import Response

from adhocracy4.api.mixins import ContentTypeMixin
from adhocracy4.api.permissions import ViewSetRulesPermission
from adhocracy4.comments.models import Comment

from .filters import CommentCategoryFilterBackend
from .filters import CommentOrderingFilterBackend
from .filters import CustomSearchFilter
from .serializers import ThreadListSerializer
from .serializers import ThreadSerializer


class CommentSetPagination(PageNumberPagination):
    page_size = 20


class CommentViewSet(mixins.CreateModelMixin,
                     mixins.ListModelMixin,
                     mixins.RetrieveModelMixin,
                     mixins.UpdateModelMixin,
                     mixins.DestroyModelMixin,
                     ContentTypeMixin,
                     viewsets.GenericViewSet):

    permission_classes = (ViewSetRulesPermission,)
    filter_backends = (filters.DjangoFilterBackend,
                       CommentOrderingFilterBackend,
                       CommentCategoryFilterBackend,
                       CustomSearchFilter)
    filter_fields = ('object_pk', 'content_type')
    ordering = ('-created')
    content_type_filter = settings.A4_COMMENTABLES
    pagination_class = CommentSetPagination
    search_fields = ('comment', '=creator__firstname', '=creator__lastname')

    def list(self, request, content_type=None, object_pk=None):
        queryset = self.filter_queryset(self.get_queryset())
        page = self.paginate_queryset(queryset)
        if page is not None:
            serializer = self.get_serializer(page, many=True)
            response = self.get_paginated_response(serializer.data)
            if 'commentID' in request.query_params:
                try:
                    commentID = int(request.query_params['commentID'])
                    if queryset.filter(id=commentID).exists():
                        response.data['comment_found'] = True
                    else:
                        parent = [item for item in
                                  queryset.values_list('id', 'child_comments')
                                  if item[1] == commentID]
                        if parent:
                            response.data['comment_found'] = True
                            response.data['comment_parent'] = parent[0][0]
                        else:
                            response.data['comment_found'] = False
                except ValueError:
                    return Response(status=status.HTTP_400_BAD_REQUEST)
            return response
        serializer = self.get_serializer(queryset, many=True)
        return Response(serializer.data)

    def get_serializer_class(self):
        if self.action == 'list':
            return ThreadListSerializer
        return ThreadSerializer

    def perform_create(self, serializer):
        serializer.save(
            content_object=self.content_object,
            creator=self.request.user
        )

    def get_permission_object(self):
        return self.content_object

    def get_queryset(self):
        child_comment_content_type_id = \
            ContentType.objects.get_for_model(Comment)
        comments = Comment.objects\
            .filter(object_pk=self.object_pk)\
            .filter(content_type_id=self.content_type.pk)
        if self.action == 'list':
            return comments\
                .exclude(content_type_id=child_comment_content_type_id)\
                .order_by('-created')
        return comments

    @property
    def rules_method_map(self):
        return ViewSetRulesPermission.default_rules_method_map._replace(
            POST='{app_label}.comment_{model}'.format(
                app_label=self.content_type.app_label,
                model=self.content_type.model
            )
        )

    def destroy(self, request, *args, **kwargs):
        comment = self.get_object()
        if self.request.user == comment.creator:
            comment.is_removed = True
        else:
            comment.is_censored = True
        comment.save()
        serializer = self.get_serializer(comment)
        return Response(serializer.data)
