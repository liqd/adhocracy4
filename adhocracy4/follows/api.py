from django_filters import rest_framework as filters
from rest_framework import mixins, permissions, viewsets

from adhocracy4.api.mixins import AllowPUTAsCreateMixin

from . import models
from .serializers import FollowSerializer


class FollowViewSet(AllowPUTAsCreateMixin,
                    mixins.ListModelMixin,
                    mixins.RetrieveModelMixin,
                    mixins.UpdateModelMixin,
                    viewsets.GenericViewSet):

    lookup_field = 'project__slug'
    queryset = models.Follow.objects
    serializer_class = FollowSerializer
    permission_classes = (permissions.IsAuthenticated,)
    filter_backends = (filters.DjangoFilterBackend,)
    filter_fields = ('enabled', )

    def get_queryset(self):
        return self.queryset.filter(creator=self.request.user)
