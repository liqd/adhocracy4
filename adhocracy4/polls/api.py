from django.db import transaction
from django.shortcuts import get_object_or_404
from rest_framework import mixins
from rest_framework import status
from rest_framework import viewsets
from rest_framework.exceptions import ValidationError
from rest_framework.response import Response

from adhocracy4.api.permissions import ViewSetRulesPermission

from .models import Choice
from .models import Poll
from .models import Question
from .models import Vote
from .serializers import PollSerializer
from .serializers import QuestionSerializer
from .validators import choice_belongs_to_question


class PollViewSet(mixins.UpdateModelMixin,
                  viewsets.GenericViewSet):
    queryset = Poll.objects.all()
    serializer_class = PollSerializer
    permission_classes = (ViewSetRulesPermission,)

    def get_permission_object(self):
        poll = self.get_object()
        return poll.module


class VoteViewSet(viewsets.ViewSet):

    permission_classes = (ViewSetRulesPermission,)

    def create(self, request, *args, **kwargs):
        choices = self.get_data(request)

        self.validate_choices(choices)

        with transaction.atomic():
            self.clear_current_choices()
            for choice in choices:
                Vote.objects.create(
                    choice_id=choice.id,
                    creator=self.request.user
                )

        question_serializer = self.get_question_serializer()
        return Response({'question': question_serializer.data},
                        status=status.HTTP_201_CREATED)

    def get_data(self, request):
        try:
            choices = [get_object_or_404(Choice, pk=choice_pk)
                       for choice_pk
                       in request.data['choices']]
        except (ValueError, AttributeError):
            raise ValidationError()

        return choices

    def validate_choices(self, choices):
        question = self.question

        if len(choices) > len(set(choices)):
            raise ValidationError('duplicate choices detected')

        if len(choices) == 0:
            raise ValidationError('empty choices detected')
        elif len(choices) > 1 and not question.multiple_choice:
            raise ValidationError('multiple choice disabled for question')

        for choice in choices:
            choice_belongs_to_question(choice, question.pk)

    def clear_current_choices(self):
        Vote.objects\
            .filter(choice__question_id=self.question.id,
                    creator=self.request.user)\
            .delete()

    @property
    def question(self):
        return get_object_or_404(
            Question,
            pk=self.kwargs['question_pk']
        )

    def get_question_serializer(self):
        question = Question.objects\
            .annotate_vote_count()\
            .get(pk=self.kwargs['question_pk'])

        return QuestionSerializer(
            question,
            context={
                'request': self.request
            }
        )

    def get_queryset(self):
        return Vote.objects.filter(
            choice__question_id=self.kwargs['question_pk'])

    def get_permission_object(self):
        return self.question.poll.module
