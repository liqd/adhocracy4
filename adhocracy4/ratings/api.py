from django.conf import settings
from django_filters import rest_framework as filters
from rest_framework import mixins
from rest_framework import viewsets
from rest_framework.response import Response
from rest_framework.serializers import ValidationError

from adhocracy4.api.mixins import ContentTypeMixin
from adhocracy4.api.permissions import ViewSetRulesPermission

from .models import Rating
from .serializers import RatingSerializer


class RatingViewSet(mixins.CreateModelMixin,
                    mixins.UpdateModelMixin,
                    ContentTypeMixin,
                    viewsets.GenericViewSet):

    queryset = Rating.objects.all()
    serializer_class = RatingSerializer
    permission_classes = (ViewSetRulesPermission,)
    filter_backends = (filters.DjangoFilterBackend,)
    filterset_fields = ('object_pk', 'content_type')
    content_type_filter = settings.A4_RATEABLES

    def perform_create(self, serializer):
        queryset = Rating.objects.filter(content_type_id=self.content_type.pk,
                                         creator=self.request.user,
                                         object_pk=self.content_object.pk)
        if queryset.exists():
            raise ValidationError(queryset[0].pk)
        serializer.save(
            content_object=self.content_object,
            creator=self.request.user
        )

    def get_permission_object(self):
        return self.content_object

    @property
    def rules_method_map(self):
        return ViewSetRulesPermission.default_rules_method_map._replace(
            POST='{app_label}.rate_{model}'.format(
                app_label=self.content_type.app_label,
                model=self.content_type.model
            )
        )

    def destroy(self, request, content_type, object_pk, pk=None):
        """
        Sets value to zero
        NOTE: Rating is NOT deleted.
        """
        rating = self.get_object()
        rating.update(0)
        serializer = self.get_serializer(rating)
        return Response(serializer.data)
