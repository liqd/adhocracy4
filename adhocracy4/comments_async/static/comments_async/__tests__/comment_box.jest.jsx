import React from 'react'
import { render, screen, fireEvent, waitFor } from '@testing-library/react'
import '@testing-library/jest-dom'
import CommentBox from '../comment_box'
import api from '../../../../static/api'

jest.mock('../../../../static/api')

afterEach(() => {
  jest.clearAllMocks()
})

describe('CommentBox Component', () => {
  const defaultProps = {
    anchoredCommentId: null,
    id: 0,
    noControlBar: false,
    subjectId: 0,
    subjectType: 108,
    useModeratorMarked: null,
    withCategories: false
  }

  test('renders CommentBox', () => {
    api.comments.setComments({
      results: []
    })
    render(<CommentBox {...defaultProps} />)
  })

  test('comments are fetched and loading spinners is hidden', () => {
    api.comments.setComments({ results: [] })
    const tree = render(<CommentBox {...defaultProps} />)
    expect(tree).toMatchSnapshot()

    expect(api.comments.get).toHaveBeenCalledTimes(1)
    const loading = screen.getByText(/Loading.../)
    expect(loading.closest('div')).toHaveClass('d-none')
  })

  test('more comments are fetched on scroll', async () => {
    global.fetch = jest.fn().mockResolvedValue({
      json: () => {
        return { results: [] }
      }
    })
    api.comments.setComments({
      results: [],
      next: 'https://liqd.net/next_comments'
    })
    render(<CommentBox {...defaultProps} />)
    expect(api.comments.get).toHaveBeenCalledTimes(1)
    const loading = screen.getByText(/Loading.../)
    expect(loading.closest('div')).toHaveClass('d-none')
    fireEvent.scroll(window, { y: 500 })
    await waitFor(() => {
      expect(fetch).toHaveBeenCalledTimes(1)
      expect(fetch).toHaveBeenCalledWith('https://liqd.net/next_comments')
    })
  })
})
