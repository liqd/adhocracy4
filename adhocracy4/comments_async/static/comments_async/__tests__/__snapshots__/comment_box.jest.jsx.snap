// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentBox Component comments are fetched and loading spinners is hidden 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section>
        <div
          class="a4-comments__commentbox__form"
        >
          <div
            class="a4-comments__alert"
          >
            The currently active phase doesn't allow to comment.
          </div>
        </div>
        <div>
          <h3
            class="a4-comments__commentbox__subtitle"
          >
            Discussion
          </h3>
          <div
            class="a4-comments__comment-control-bar"
          >
            <div
              class="form-group form-group--inline"
            >
              <div
                class="form-group"
              >
                <form
                  class="a4-control-bar__search__form"
                >
                  <div
                    class="form-group"
                  >
                    <label
                      class="form-label"
                      for="searchterm"
                    >
                      Search
                    </label>
                    <div
                      class="a4-control-bar__search__term"
                    >
                      <div
                        class="input-wrapper"
                      >
                        <i
                          aria-hidden="true"
                          class="a4-control-bar__search__logo-start"
                        />
                        <input
                          class="form-control a4-control-bar__search__input"
                          id="searchterm"
                          placeholder=""
                          type="search"
                          value=""
                        />
                      </div>
                      <button
                        class="a4-control-bar__search__input-submit"
                        title="Start search"
                        type="submit"
                      >
                        <span
                          class="a4-sr-only"
                        >
                          Start search
                        </span>
                        <i
                          aria-hidden="true"
                          class="a4-control-bar__search__logo-end"
                        />
                      </button>
                    </div>
                  </div>
                </form>
              </div>
              <div
                class="form-group"
              >
                <div
                  class="a4-control-bar__sorting"
                >
                  <div
                    class="form-group a4-forms__select"
                  >
                    <label
                      class="form-label a4-forms__select__label"
                      for="id_ordering"
                    >
                      Ordering
                    </label>
                    <div
                      class="a4-forms__select__wrapper"
                    >
                      <select
                        class="form-control a4-forms__select__input"
                        id="id_ordering"
                      >
                        <option
                          value="new"
                        >
                          Newest
                        </option>
                        <option
                          value="pos"
                        >
                          Most up votes
                        </option>
                        <option
                          value="neg"
                        >
                          Most down votes
                        </option>
                        <option
                          value="ans"
                        >
                          Most answers
                        </option>
                        <option
                          value="dis"
                        >
                          Last discussed
                        </option>
                      </select>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <ul
          class="u-list-reset a4-comments"
        />
        <div
          class="d-none"
        >
          <i
            aria-hidden="true"
            class="fa fa-spinner fa-pulse"
          />
          <span
            class="a4-sr-only"
          >
            Loading...
          </span>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section>
      <div
        class="a4-comments__commentbox__form"
      >
        <div
          class="a4-comments__alert"
        >
          The currently active phase doesn't allow to comment.
        </div>
      </div>
      <div>
        <h3
          class="a4-comments__commentbox__subtitle"
        >
          Discussion
        </h3>
        <div
          class="a4-comments__comment-control-bar"
        >
          <div
            class="form-group form-group--inline"
          >
            <div
              class="form-group"
            >
              <form
                class="a4-control-bar__search__form"
              >
                <div
                  class="form-group"
                >
                  <label
                    class="form-label"
                    for="searchterm"
                  >
                    Search
                  </label>
                  <div
                    class="a4-control-bar__search__term"
                  >
                    <div
                      class="input-wrapper"
                    >
                      <i
                        aria-hidden="true"
                        class="a4-control-bar__search__logo-start"
                      />
                      <input
                        class="form-control a4-control-bar__search__input"
                        id="searchterm"
                        placeholder=""
                        type="search"
                        value=""
                      />
                    </div>
                    <button
                      class="a4-control-bar__search__input-submit"
                      title="Start search"
                      type="submit"
                    >
                      <span
                        class="a4-sr-only"
                      >
                        Start search
                      </span>
                      <i
                        aria-hidden="true"
                        class="a4-control-bar__search__logo-end"
                      />
                    </button>
                  </div>
                </div>
              </form>
            </div>
            <div
              class="form-group"
            >
              <div
                class="a4-control-bar__sorting"
              >
                <div
                  class="form-group a4-forms__select"
                >
                  <label
                    class="form-label a4-forms__select__label"
                    for="id_ordering"
                  >
                    Ordering
                  </label>
                  <div
                    class="a4-forms__select__wrapper"
                  >
                    <select
                      class="form-control a4-forms__select__input"
                      id="id_ordering"
                    >
                      <option
                        value="new"
                      >
                        Newest
                      </option>
                      <option
                        value="pos"
                      >
                        Most up votes
                      </option>
                      <option
                        value="neg"
                      >
                        Most down votes
                      </option>
                      <option
                        value="ans"
                      >
                        Most answers
                      </option>
                      <option
                        value="dis"
                      >
                        Last discussed
                      </option>
                    </select>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <ul
        class="u-list-reset a4-comments"
      />
      <div
        class="d-none"
      >
        <i
          aria-hidden="true"
          class="fa fa-spinner fa-pulse"
        />
        <span
          class="a4-sr-only"
        >
          Loading...
        </span>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
